XO.py

```diff
@@ -0,0 +1,157 @@
+from random import shuffle
+
+# size of board
+N = 3
+
+# Utility functions
+
+
+def print_board(board):
+    print()
+    print('Board:')
+    for line_ in board:
+        for item in line_:
+            print(item, end='')
+        print()
+    print()
+
+
+def check_input_validity(input_type, curr_player):
+    not_valid = True
+    while not_valid:
+        print('{}, please insert the {} number'.format(curr_player, input_type))
+        try:
+            input_value = int(input())
+        except:
+            print('{}, please insert a valid {} number'.format(curr_player, input_type))
+            continue
+        if input_value < 0 or input_value >= N:
+            print('{}, please insert a valid {} number'.format(curr_player, input_type))
+            continue
+        not_valid = False
+    return input_value
+
+
+def check_line(board, line, curr_player, curr_player_symbol, is_win):
+    for sym in board[line]:
+        if sym != curr_player_symbol:
+            is_win = False
+            break
+    if is_win:
+        return declare_win(curr_player)
+
+
+def check_col(board, col, curr_player, curr_player_symbol, is_win):
+    for line in board:
+        if line[col] != curr_player_symbol:
+            is_win = False
+            break
+    if is_win:
+        return declare_win(curr_player)
+
+
+def check_main_diagonal(board, line, col, curr_player, curr_player_symbol, is_win):
+    if line == col:

```

Comment:
This check is misleading, because when looking at the function's name, it seems like it'll always check the main diagonal (which is good). If you want to optimize, it's great, but it's better to do it outside the scope of the check function.

Reply:
Done. In this case I removed the optimization altogether to make the code more simple and readable.